File: frontend/src/types/media.ts
export interface Media {
  id: string;
  title: string;
  original_name: string;
  year: number;
  quality: string;
  genre: string;
  file_path: string;
  file_size: number;
  duration: number;
  mime_type: string;
  thumbnail_path: string;
  status: 'pending' | 'downloading' | 'processing' | 'ready' | 'error';
  source: 'torrent' | 'upload' | 'scan';
  language: string;
  last_position?: number;
  default_start_time?: number;
  created_at: string;
  updated_at: string;
}

export interface Category {
  id: string;
  name: string;
  created_at: string;
}

export interface Clip {
  id: string;
  media_id: string;
  title: string;
  start_time: number;
  end_time: number;
  thumbnail_path: string;
  category_ids?: string[];
  categories?: Category[];
  created_at: string;
  updated_at: string;
}

export interface CreateClipPayload {
  media_id: string;
  title: string;
  start_time: number;
  end_time: number;
  category_ids?: string[];
  thumbnail_frame_time?: number;
}

export interface LibraryStats {
  count: number;
  total_size: number;
}

export interface TorrentStatus {
  media_id: string;
  title: string;
  status: string;
  total_bytes: number;
  completed_bytes: number;
  progress_pct: number;
  download_rate_bps: number;
  peers: number;
}

export interface TorrentTarget {
  title: string;
  size: string;
  link: string;
}

export interface Preference {
  key: string;
  value: string;
}

export interface DirectoryItem {
  name: string;
  path: string;
}

export interface BrowseData {
  current_path: string;
  parent_path: string;
  directories: DirectoryItem[];
}

export interface Download {
  id: string;
  title: string;
  status: 'downloading' | 'completed' | 'failed' | 'cancelled';
  type: 'torrent' | 'youtube';
  progress: number;
  total_size: number;
  completed_size: number;
  download_speed: number;
  eta: string;
  dest_path: string;
  created_at: string;
  updated_at: string;
}

File: backend/internal/model/media.go
package model

import (
	"errors"
	"strings"
	"time"
)

// MediaStatus represents the state of a media file ingest pipeline.
type MediaStatus string

const (
	StatusPending     MediaStatus = "pending"
	StatusDownloading MediaStatus = "downloading"
	StatusProcessing  MediaStatus = "processing"
	StatusReady       MediaStatus = "ready"
	StatusError       MediaStatus = "error"
)

// MediaSource tells us how this movie was added.
type MediaSource string

const (
	SourceTorrent MediaSource = "torrent"
	SourceUpload  MediaSource = "upload"
	SourceScan    MediaSource = "scan"
	SourceYoutube MediaSource = "youtube"
)

// Media is the central domain struct representing a movie in our database.
type Media struct {
	ID            string      `json:"id"`
	Title         string      `json:"title"`
	OriginalName  string      `json:"original_name"`
	Year          int         `json:"year"`
	Quality       string      `json:"quality"`
	Genre         string      `json:"genre"`
	FilePath      string      `json:"file_path"`
	FileSize      int64       `json:"file_size"`
	Duration      int         `json:"duration"` // in seconds
	MimeType      string      `json:"mime_type"`
	ThumbnailPath string      `json:"thumbnail_path"`
	Status        MediaStatus `json:"status"`
	Source        MediaSource `json:"source"`
	Language         string      `json:"language"`           // e.g. "en", "hi"
	LastPosition     int         `json:"last_position"`      // last watched position in seconds
	DefaultStartTime int         `json:"default_start_time"` // default auto-playing start position in seconds
	CreatedAt        time.Time   `json:"created_at"`
	UpdatedAt        time.Time   `json:"updated_at"`
}

// Validate verifies domain integrity constraints before save/update operations.
func (m *Media) Validate() error {
	if strings.TrimSpace(m.ID) == "" {
		return errors.New("media ID is required")
	}
	if strings.TrimSpace(m.Title) == "" {
		return errors.New("media title is required")
	}
	if strings.TrimSpace(m.FilePath) == "" {
		return errors.New("media file path is required")
	}
	if m.Status == "" {
		m.Status = StatusPending
	}
	if m.Source == "" {
		m.Source = SourceScan
	}
	if strings.TrimSpace(m.Language) == "" {
		m.Language = "en"
	}
	return nil
}
